shelley/chain-and-ledger/executable-spec/test/Generator/DCertTrace.hs

```diff
@@ -0,0 +1,154 @@
+  (genDCerts)
+  where
+
+import           Control.Monad.Trans.Reader (runReaderT)
+import           Data.Functor.Identity (runIdentity)
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map (lookup)
+import           Data.Maybe (catMaybes)
+import           Data.Sequence (Seq)
+import qualified Data.Sequence as Seq
+import           Data.Word (Word64)
+import           Lens.Micro ((^.))
+import           Numeric.Natural (Natural)
+import           Test.QuickCheck (Gen)
+
+import           BaseTypes (ShelleyBase)
+import           BaseTypes (Globals)
+import           Coin (Coin)
+import           ConcreteCryptoTypes (AnyKeyHash, DCert, DELPL, DPState, KeyPair)
+import           Control.State.Transition (BaseM, Embed, Environment, PredicateFailure, STS, Signal,
+                     State, TRC (..), TransitionRule, initialRules, judgmentContext, trans,
+                     transitionRules, wrapFailed)
+import           Control.State.Transition.Trace (TraceOrder (OldestFirst), traceSignals)
+import qualified Control.State.Transition.Trace.Generator.QuickCheck as QC
+import           Delegation.Certificates (decayKey, isDeRegKey)
+import           Generator.Core.Constants (maxCertsPerTx, numBaseScripts)
+import           Generator.Core.QuickCheck (coreNodeKeys, traceKeyPairs, traceKeyPairsByStakeHash,
+                     traceMSigCombinations, traceMSigScripts, traceVRFKeyPairs)
+import           Generator.Delegation.QuickCheck (CertCred (..), genDCert)
+import           GHC.Generics (Generic)
+import           LedgerState (dstate, keyRefund, stkCreds, _pstate, _stPools)
+import           PParams (PParams)
+import           Slot (SlotNo (..))
+import           STS.Delpl (DelplEnv (..))
+import           Tx (getKeyCombination)
+import           TxData (Ix, Ptr (..))
+import           UTxO (totalDeposits)
+
+import           Test.Utils (testGlobals)
+
+-- | This is a non-spec STS used to generate a sequence of certificates with
+-- witnesses.
+data CERTS
+
+instance STS CERTS where
+  type Environment CERTS = (SlotNo, Ix, PParams, Coin)
+  type State CERTS = (DPState, Ix)
+  type Signal CERTS = Maybe (DCert, CertCred)
+
+  type BaseM CERTS = ShelleyBase
+
+  data PredicateFailure CERTS
+    = CertsFailure (PredicateFailure DELPL)
+    deriving (Show, Eq, Generic)
+
+  initialRules = []
+  transitionRules = [certsTransition]
+
+certsTransition :: TransitionRule CERTS
+certsTransition = do
+  TRC ( (slot, txIx, pp, reserves)
+      , (dpState, nextCertIx)
+      , c) <- judgmentContext
+
+  case c of
+    Just (cert, _wits) -> do
+      let ptr = Ptr slot txIx nextCertIx
+      dpState' <- trans @DELPL $ TRC (DelplEnv slot ptr pp reserves, dpState, cert)
+
+      pure $ (dpState', nextCertIx + 1)
+    Nothing ->
+      pure $ (dpState, nextCertIx)
+
+instance Embed DELPL CERTS where
+  wrapFailed = CertsFailure
+
+instance QC.HasTrace CERTS Word64 where
+  envGen _ = undefined -- we won't be using this
```

Comment:
maybe make this an `error` with an easy to identify message in case someone does use this?